### PR TITLE
Gradle init script improvements

### PR DESCRIPTION
**Pull request by erichaagdev**
## ❄️ Winter Update PR Navigator ❄️

1. #261
2. #262
3. #263
4. ⛄ #264
5. #265
6. #266
7. #267
8. #272
9. #274 
10. #283

⛄ = you are here

## Summary

This PR is a rebased version of #121 from @bigdaz. I decided to create a new branch and PR in order to preserve his branch as it. That way I could easily switch between the two as I verified these changes. Once this PR is merged #121 can be closed.

|   |
| --- |
| ⚠️ While testing these changes, I discovered a bug that exists in the #121 version when running with Gradle 5.x. When running the scripts with `-e` both `com.gradle.build-scan` and `com.gradle.enterprise` are applied. This can be seen in this [Build Scan](https://scans.gradle.com/s/a7dz6nhqskzne/build-dependencies?toggled=W1swXSxbMCwwXSxbMV0sWzEsMF1d). It seems this scenario [causes an error when applying CCUD](https://scans.gradle.com/s/a7dz6nhqskzne/failure?focused-stack-frame=0-1-2-3#1).<br><br>I was able to resolve this by removing `--scan` as a parameter to the build validation scripts' Gradle invocation. Since the initialization scripts take care of configuring and setting up Gradle Enterprise I don't think the `--scan` parameter is required.<br><br>However, `--scan` is still displayed in the printed Gradle invocation command. This is somewhat misleading now. In my opinion, it was  also misleading before since the displayed command does not include the initialization scripts that perform additional configuration transparent to the user.  |

The only other change from #121 is that I removed all `logger` calls in the initialization scripts since they created unnecessary noise in the validation scripts output that was not present before. 

## Testing

In order to test this I ran the following experiments and verified a successful outcome. 

The following parameter combinations were tested with Gradle `5.x`, `6.x`, and `7.x`:

- online
- offline (`-x`)
- online, enable GE (`-e`)
- offline (`-x`), enable GE  (`-e`)

```sh
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/5.x/ge -t 'build'
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/5.x/ge -t 'build' -x
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/5.x/no-ge -t 'build' -e -s https://ge.solutions-team.gradle.com
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/5.x/no-ge -t 'build' -e -x
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/6.x/ge -t 'build'
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/6.x/ge -t 'build' -x
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/6.x/no-ge -t 'build' -e -s https://ge.solutions-team.gradle.com
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/6.x/no-ge -t 'build' -e -x
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/7.x/ge -t 'build'
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/7.x/ge -t 'build' -x
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/7.x/no-ge -t 'build' -e -s https://ge.solutions-team.gradle.com
./03-validate-local-build-caching-different-locations.sh -r git@github.com:gradle/ge-solutions.git -p sample-projects/gradle/7.x/no-ge -t 'build' -e -x
./04-validate-remote-build-caching-ci-ci.sh -1 https://ge.solutions-team.gradle.com/s/fvwne3ht4xgjg -2 https://ge.solutions-team.gradle.com/s/tjiwbi2b7oioo
./05-validate-remote-build-caching-ci-local.sh -1 https://ge.solutions-team.gradle.com/s/lc73bw56kdkxi
```

### CODE REVIEW

**Comment by jthurne** [comment]
> However, --scan is still displayed in the printed Gradle invocation command. This is somewhat misleading now. In my opinion, it was also misleading before since the displayed command does not include the initialization scripts that perform additional configuration transparent to the user.

The intent with the displayed command is to show the user a simple command they can use themselves when running Gradle directly to achieve a similar result. With that in mind, we don't print out the full command used so as not to distract or misslead the user.

These initialization scripts have become sufficiently complex that I think many users would have difficulty interpreting them or understanding what they do.  Which is fine, but it's a reason why we don't necessarily want to display their use in the command that we execute.

File: components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle [hunk 1]
```diff
@@ -1,69 +1,160 @@
 import org.gradle.util.GradleVersion
 import java.nio.charset.StandardCharsets;
 
+if (GradleVersion.current() < GradleVersion.version("5.0")) {
+    throw new IllegalStateException("Build validation not supported for Gradle ${GradleVersion.current()}. Upgrade your project's build to Gradle 5 or newer.")
```
Comment (jthurne) [inline components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle:5]:
```suggestion
    throw new IllegalStateException("The Gradle Enterprise Build Validation Scripts are not supported with Gradle ${GradleVersion.current()}. Please upgrade your project's build to Gradle 5 or newer.")
```

File: components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle [hunk 2]
```diff
@@ -1,69 +1,160 @@
...
+        }
+        def errorFile = new File(experimentDir, "build-scan-publish-error.txt")
+        buildScan.onError { error ->
+            errorFile.text = error
+        }
+    }
+}
 
-    clearWarnings()
+// register buildScanPublished listener and optionally apply the GE / Build Scan plugin
```
Comment (jthurne) [inline components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle:75]: -1: I don't see a buildScanPublished listener being registered below anywhere.

```suggestion
// Optionally apply the GE / Build Scan plugin
```

File: components/scripts/lib/gradle.sh [hunk 3]
```diff
@@ -10,20 +10,17 @@ invoke_gradle() {
     cd "${project_dir}" > /dev/null 2>&1 || die "ERROR: The subdirectory ${project_dir} (set with --project-dir) does not exist in ${project_name}." "${INVALID_INPUT}"
   fi
 
+  args+=(--init-script "${INIT_SCRIPTS_DIR}/configure-gradle-enterprise.gradle")
+
   if [ "$enable_ge" == "on" ]; then
-    args+=(--init-script "${INIT_SCRIPTS_DIR}/enable-gradle-enterprise.gradle")
+    args+=("-Dcom.gradle.enterprise.build_validation.enableGradleEnterprise=true")
```
Comment (jthurne) [inline components/scripts/lib/gradle.sh:16]: 0: I'm curious, why are we setting these as system properties instead of as project properties?

**Comment by jthurne** [comment]
> Once this PR is merged #121 can be closed.

Would it be okay if we closed it now?  I've gotten confused a few times between #121 and this PR.  Closing a PR doesn't delete the underlying branch, so you could still compare your branch against Daz's original branch.

File: components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle (hunk 2 above)
Comment (erichaagdev) [reply components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle:75]: The listener is being applied [here](https://github.com/gradle/gradle-enterprise-build-validation-scripts/blob/a2e2bf985267d51e5b2c59502c72bba7da28e811/components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle#L107) and [here](https://github.com/gradle/gradle-enterprise-build-validation-scripts/blob/a2e2bf985267d51e5b2c59502c72bba7da28e811/components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle#L139).

File: components/scripts/lib/gradle.sh (hunk 3 above)
Comment (erichaagdev) [reply components/scripts/lib/gradle.sh:16]: That's a great question. I am not sure. I only continued the pattern started by @bigdaz in #121. 

Can you think of any reason not to use a project property?

File: components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle (hunk 2 above)
Comment (jthurne) [reply components/scripts/gradle/gradle-init-scripts/configure-gradle-enterprise.gradle:75]: So it is! 👍

**Comment by erichaagdev** [comment]
A positive side effect of removing `--scan` is that if a build is not actually configured with Gradle Enterprise, then it will not try to publish a Build Scan to scans.gradle.com. For example:

```
./03-validate-local-build-caching-different-locations.sh -r https://github.com/android/nowinandroid -t assembleDemoDebug
```

If we run the above without these changes, the build will fail with the following error:

image

As you can see, it is asking to accept the terms of service so that a Build Scan may be published to scans.gradle.com. This is risky when running the scripts for a project whose Build Scan should not be made publicly available. With these changes, the user will see the following instead:

image

File: components/scripts/lib/gradle.sh (hunk 3 above)
Comment (bigdaz) [reply components/scripts/lib/gradle.sh:16]: It was my understanding that project properties aren't available when an init script evaluates. But I could well be mistaken.

Comment (erichaagdev) [reply components/scripts/lib/gradle.sh:16]: That makes sense @bigdaz. I'm having trouble finding it anywhere in the documentation, but I did make a small sample project to test.

```
* What went wrong:
A problem occurred evaluating initialization script.
> The default project is not yet available for build.
```

image

**Pull request merged**